### DISCUSSION TITLE
Customization feature idea

### DISCUSSION THREAD

**Discussion by VirtualPedestrian** [EmiyaSyahriel/CrossLauncher#5] [Ideas]
I love the launcher but it lack customization feature. Here's some idea for customization feature:
Customizable icon
Hide app
Re-arrange-able column
App preview backround
Ability to change font
Hide notification/status bar
Shortcut function

**Comment by EmiyaSyahriel** [comment]
Customizable icon, app hiding, app background (PIC0.png / PIC1.png), and possibly audio theming (like how PS3 XMB does) is already on my mind. And hiding system bars should be simple to implement. Custom fonts and arrangable columns might be a problem, but I might be able to made one.

Now, I'd like to ask your opinion about how would custom resources (like icons, audio, cover and fonts) to be loaded. The way I currently have in mind is by simulating a PS3 File System (specifically dev_hdd0) inside the app's persistent directory (`"/(media storage/sdcard root)/Android/data/id.psw.crosslauncher/files/"`) and then load these resources from them. User may modify the files inside manually... or by handling the opening of a special file type (a PKG-esque file type maybe?)... or maybe using a description page where you can edit the icon and cover? The third is actually a bit troublesome to implement, but is of course doable.

The current file system loader is already implemented at [VshCustomizationExtension.kt](https://github.com/EmiyaSyahriel/CrossLauncher/blob/9afe3ddabdd62ca8406a2e9c153ec58c905def34/app/src/main/java/id/psw/vshlauncher/VshCustomizationExtension.kt) ( if you can read it ). File structure reference : [PS3 Dev Wiki / Content Information Files](https://www.psdevwiki.com/ps3/Content_Information_Files)

**Reply by VirtualPedestrian** [reply]
The first one

**Reply by EmiyaSyahriel** [reply]
Got it, I have some free time right now. I'm currently working the subcontent (item foldering) and the video player input, maybe I'll release this customization feature (at least, functional) in 3 or 4 release after this (if not less).

**Reply by VirtualPedestrian** [reply]
Can you build an updated apk, btw do you know how to compile apk without Android studio?

**Reply by EmiyaSyahriel** [reply]
I can, But the app setting section is currently unusable, Idk if you still want to use it without being able to change any setting.

AFAIK, compiling an APK only need Android SDK with gradle (and it's all command line), which is bundled with Android Studio but can be downloaded separately, and is available for Windows, Linux and macOS. There is a StackOverflow answer about it [here](https://stackoverflow.com/a/55393934). But if what you mean is building APK from Android itself, All I know is only to recompile decompiled APK and it cannot be used with source code, CMIIW.

**Reply by VirtualPedestrian** [reply]
You can change the app icon manually right? If so I'd like the updated apk

**Reply by EmiyaSyahriel** [reply]
Here : https://drive.google.com/file/d/16-kiizZ7GukjHJVjkZZu-1DCXTbGhf4A/view?usp=sharing

I haven't tested some customization feature, It's just that you don't want to reset any setting as you cannot access them for now.

**Reply by VirtualPedestrian** [reply]
How do you change icon?

**Reply by EmiyaSyahriel** [reply]
You can just drop an icon named `ICON0.PNG` at `(Device Media Storage)/Android/data/id.psw.vshlauncher/files/dev_hdd0/games/(App Package Name)/` (Make these folders if not yet there).

That should do it, I'm not sure if I'm doing it right or not as I haven't working on it or even update the codes in github repo for quite a while now. *Unreal engine's C++ is just too complex for sure 😥*

**Reply by VirtualPedestrian** [reply]
Doesn't work for me

**Reply by PSPhoneLemming** [reply]
Woops! ive sent a discussion thread about app hiding, my bad. i am too dumb to talk with cool developers like you, emiya

**Comment by EmiyaSyahriel** [comment]
Ah yes, tried the build today. Looks like I have to rework the launcher's entire icon system so it can be more modular, customizable, can act as a container and API-exportable with lesser, manageable code.

The previous icon data system is only using limited hierarchy with 2D indexing, and the code is bloated with fixed functions for each condition when the icon have subcontent and not, and looks like I also overlooked some Java File API to load custom resource files and using the hard way to load them. Now I'm planning on full hierarchy icon data system for the icons so it can be nested each other with the same render and interaction code.

And also, I know that the "PS3 File System"-ish Folder hierarchy for customization file paths is troublesome to navigate with Android file explorer apps. but don't worry, I might also made this paths customizable too sometime in future.

**Reply by VirtualPedestrian** [reply]
I noticed that there's 2 file named XMBWaveRenderer.kt and XMBWaveService.kt. Are you planning to add the wave wallpaper used in PS3 and PSP?

**Reply by EmiyaSyahriel** [reply]
Yes, But not soon (or maybe never), Those files currently serve only as placeholder.
There is already some way to use XMB Wave as Android Live Wallpaper, would it be rendered in realtime or non-realtime (e.g. video).

**Reply by VirtualPedestrian** [reply]
There's a live wallpaper that's very accurate to PSP wallpaper. You could somehow extract the app's source code and use it on your app. You can't change the wallpaper color for some reason though. https://forum.xda-developers.com/t/apk-xperia-play-live-wallpapers-classic-and-original.1025087/

**Reply by EmiyaSyahriel** [reply]
I can't find any source code (or links to it) in that post. I might made my own based on sources around GitHub or maybe the whole internet though. Also, It might be a nice opportunity for me to learn some OpenGL stuff on Android too, Since until today I'm just throwing things over Unity and Unreal to render it.

Edit: I downloaded the app, examine the content of it and nope, The entire rendering is handled inside the Native part of the app while the JVM part (Android's own runtime) is just for Android OS's interaction. I cannot easily decompile it, And should be forbidden too, it's a Sony Ericsson official live wallpaper.

Anyway, that's a nice reference so I don't need to keep my PS3 turned on all the day to only watch it's wave 🙂. I still have to fix the launcher part though 😞

**Reply by Adorno84** [reply]
Hello I hope you are doing well, I love your project a lot of potential here. I am a layman so forgive me if this is a silly suggestion but have you considered copying the way that retroarch does their XMB background? I believe this is the source code for it right here: https://github.com/libretro/RetroArch/blob/efe34f52afed900cd027e78f61ce4ef5f3bea763/menu/drivers/xmb.c

heres an older video of it in action I believe they have added many features since then :
https://www.youtube.com/watch?v=vgqSfo3qML4

Sorry if this doesnt help like i said I am a layman when it comes to programming.

**Reply by VirtualPedestrian** [reply]
> Hello I hope you are doing well, I love your project a lot of potential here. I am a layman so forgive me if this is a silly suggestion but have you considered copying the way that retroarch does their XMB background? I believe this is the source code for it right here: https://github.com/libretro/RetroArch/blob/efe34f52afed900cd027e78f61ce4ef5f3bea763/menu/drivers/xmb.c
> 
> heres an older video of it in action I believe they have added many features since then : https://www.youtube.com/watch?v=vgqSfo3qML4
> 
> Sorry if this doesnt help like i said I am a layman when it comes to programming.

Retroarch wave looks nice, but it looks terrible on low resolution 
![Screenshot_20210930-093245](https://user-images.githubusercontent.com/73787177/135384332-97c381e1-9d46-4725-aaf0-19d1ccd5dae2.png)
![Screenshot_20210930-093611](https://user-images.githubusercontent.com/73787177/135384350-44bbbf51-fdae-484e-a528-dc1998b42f05.png)

**Reply by Adorno84** [reply]
thats a bummer, how about this? https://play.google.com/store/apps/details?id=com.tbehlman.xmb

looks good on one of my phones but doesnt work with the other unfortunately

**Reply by Adorno84** [reply]
![Screenshot_20210930-010558](https://user-images.githubusercontent.com/91628119/135390818-10cb9651-b6b7-47e5-8f33-4a60d998a921.png)

Feels really nice together

**Reply by EmiyaSyahriel** [reply]
@Adorno84 About the Wave live wallpaper, I've already made one based off RetroArch XMB source code (Looks raw though), it's just not yet customizable and sometimes the gradient background becomes fully black. It's already in the project and you can compile it (it requires Android NDK in addition to Android SDK

![WhatsApp Image 2021-09-30 at 23 23 06](https://user-images.githubusercontent.com/24495564/135493863-240f9c69-459e-4071-b7b3-05079f11f439.jpeg)
).

**Reply by VirtualPedestrian** [reply]
You should implement PPSSPP wallpaper
![Screenshot_20211014-101755](https://user-images.githubusercontent.com/73787177/137239469-010baf0d-21d9-4553-be5e-6614e1f0f9e6.png)
![Screenshot_20211014-101748](https://user-images.githubusercontent.com/73787177/137239476-274b86af-313d-4523-af9e-d777782037b9.png)

**Reply by PSPhoneLemming** [reply]
Nice! there doesnt seem to be a good wave live wallpaper, its great seeing you add that :)
